rfc/src/rfcs/0002-function-stubbing.md

```diff
@@ -156,7 +156,45 @@ Given a set of `original`-`replacement` pairs, Kani will exit with an error if
 1. a specified `original` function/method does not exist;
 2. a specified `replacement` stub does not exist;
 3. the user specifies conflicting stubs for the same harness (e.g., if the same `original` function is mapped to multiple `replacement` functions); or
-4. the signature of the `replacement` stub is not compatible with the signature of the `original` function/method.
+4. the signature of the `replacement` stub is not compatible with the signature of the `original` function/method (see next section).
+
+### Stub compatibility and validation
+
+When considering whether a function/method can be replaced with some given stub, we want to allow some measure of flexibility, while also ensuring that we can provide the user with useful feedback if stubbing results in misformed code.
+We consider a stub to be compatible with some function/method if all the following conditions are met:
+
+- They have the same return type. The names of type variables need to match.
+- They have the same number of parameters.
+- Each parameter in the stub has the same type as the corresponding parameter in the original function/method.
+The names of type variables need to match.
+- During monomorphization, every statically dispatched call to a trait method can be resolved.
+
+The final point is the most subtle.
+We do not require that a type variable in the signature of the stub implements the same traits as the corresponding type variable in the signature of the original function/method.
+However, Kani will reject a stub if a trait mismatch leads to a situation where a statically dispatched call to a trait method cannot be resolved during monomorphization.
+For example, this restriction rules out the following harness:
+```rust
+fn foo<T>(_x: T) -> bool {
+    false
+}
+
+trait DoIt {
+    fn do_it(&self) -> bool;
+}
+
+fn bar<T: DoIt>(x: T) -> bool {
+    x.do_it()
+}
+
+#[kani::proof]
+#[kani::stub(foo, bar)]
+fn harness() {
+    assert!(foo("hello"));
+}
+```
+The call to the trait method `DoIt::do_it` is unresolvable in the stub `bar` when the type variable `T` is instantiated with the type `&str`.
+On the other hand, our approach provides some flexibility, such as allowing our earlier example of mocking randomization: both `rand::random` and `my_random` have the type `() -> T`, but in the first case `T` is restricted such that the type `Standard` implements `Distribution<T>`, whereas in the latter case `T` has to implement `kani::Arbitrary`.
+This trait mismatch is harmless because it does not result in any unresolvable behavior when `T` is instantiated with `u32`.
```

Comment:
Maybe remove this line. I think what you mean here is to say that this flexibility allow users to write expressive generic stubs with well defined behavior.

Reply:
I changed the last sentence to this:

> This trait mismatch is allowed because at this call site `T` is instantiated with `u32`, which implements `kani::Arbitrary`.